json/src/main/java/io/airlift/json/JsonSubType.java

```diff
@@ -0,0 +1,231 @@
+            this.subClassPropertyValues = ImmutableMap.copyOf(subClassPropertyValues);
+        }
+
+        @Override
+        public void findAndAddVirtualProperties(MapperConfig<?> config, AnnotatedClass ac, List<BeanPropertyWriter> properties)
+        {
+            String propertyValue = subClassPropertyValues.get(ac.getRawType());
+            if (propertyValue != null) {
+                // mostly copied from JacksonAnnotationIntrospector._constructVirtualProperty()
+
+                JavaType stringType = config.constructType(String.class);
+
+                AnnotatedMember member = new VirtualAnnotatedMember(ac, ac.getRawType(), propertyName, stringType);
+                SimpleBeanPropertyDefinition propDef = SimpleBeanPropertyDefinition.construct(config, member, PropertyName.construct(propertyName));
+                AttributePropertyWriter propertyWriter = new AttributePropertyWriter(propertyName, propDef, null, stringType)
+                {
+                    @Override
+                    protected Object value(Object bean, JsonGenerator jgen, SerializerProvider prov)
+                    {
+                        return propertyValue;
+                    }
+                };
+
+                properties.add(propertyWriter);
+            }
+        }
+
+        @Override
+        public Version version()
+        {
+            return Version.unknownVersion();
+        }
+    }
+
+    private static class Deserializer<B>
+            extends JsonDeserializer<B>
+    {
+        private final String propertyName;
+        private final Supplier<Map<String, Class<?>>> subClassPropertyValues;
+
+        public Deserializer(String propertyName, Supplier<Map<String, Class<?>>> subClassPropertyValues)
+        {
+            this.propertyName = requireNonNull(propertyName, "propertyName is null");
+            this.subClassPropertyValues = subClassPropertyValues;
+        }
+
+        @Override
+        public B deserialize(JsonParser p, DeserializationContext ctxt)
+                throws IOException
+        {
+            TreeNode treeNode = p.readValueAsTree();
+            TreeNode propertyNode = treeNode.get(propertyName);
+
+            String propertyValue = ((propertyNode != null) && propertyNode.isValueNode()) ? ((ValueNode) propertyNode).textValue() : null;
+            if (propertyValue == null) {
+                throw new IllegalArgumentException("JSON expected to have a property named %s. Double check the addBinding() or addPermittedSubClassBindings()".formatted(propertyName));
+            }
+
+            Class<?> subClass = subClassPropertyValues.get().get(propertyValue);
+            if (subClass == null) {
+                throw new IllegalArgumentException("No binding was made for property name %s and value %s. Double check the addBinding() or addPermittedSubClassBindings().".formatted(propertyName, propertyValue));
+            }
+
+            //noinspection unchecked
```

Comment:
I don't think this suppression works for `javac`. Probably need a helper method for this.